dtm0/recovery.c

```diff
@@ -0,0 +1,2170 @@
+
+	F(got) = false;
+	M0_SET0(&F(item));
+	M0_SET0(&F(op));
+	m0_be_op_init(&F(op));
+	m0_be_queue_lock(&rf->rf_eolq);
+	M0_BE_QUEUE_GET(&rf->rf_eolq, &F(op), &F(item), &F(got));
+	m0_be_queue_unlock(&rf->rf_eolq);
+	M0_CO_YIELD_RC(CO(fom), m0_be_op_tick_ret(&F(op), fom, RFS_WAITING));
+	m0_be_op_fini(&F(op));
+
+	*out = F(got) ? F(item) : (struct eolq_item) { .ei_type = EIT_END };
+}
+
+static void restore(struct m0_fom *fom,
+		    enum m0_ha_obj_state *out, bool *eoq)
+{
+	struct recovery_fom   *rf = M0_AMB(rf, fom, rf_base);
+	struct m0_dtm0_log_rec record;
+	int                    rc = 0;
+
+	M0_CO_REENTER(CO(fom),
+		      struct dtm0_req_fop redo;
+		      struct m0_fid       initiator;
+		      struct m0_be_op     reply_op;
+		      bool                next;
+		      );
+
+	recovery_machine_log_iter_init(rf->rf_m, &rf->rf_log_iter);
+
+	/* XXX: race condition in the case where we are stopping the FOM. */
+	F(initiator) = recovery_fom_local(rf->rf_m)->rf_tgt_svc;
+
+	M0_SET0(&F(reply_op));
+	m0_be_op_init(&F(reply_op));
+
+	do {
+		M0_SET0(&record);
+		rc = recovery_machine_log_iter_next(rf->rf_m, &rf->rf_log_iter,
+						    &rf->rf_tgt_svc, NULL,
+						    &record);
+		/* Any value except zero means that we should stop recovery. */
+		F(next) = rc == 0;
+		F(redo) = (struct dtm0_req_fop) {
+			.dtr_msg       = DTM_REDO,
+			.dtr_initiator = F(initiator),
+			.dtr_payload   = record.dlr_payload,
+			.dtr_txr       = record.dlr_txd,
+			.dtr_flags     = F(next) ? 0 : M0_BITS(M0_DMF_EOL),
+		};
+
+		recovery_machine_redo_post(rf->rf_m, &rf->rf_base,
+					   &rf->rf_tgt_proc, &rf->rf_tgt_svc,
+					   &F(redo), &F(reply_op));
+
+		M0_LOG(M0_DEBUG, "out-redo: (m=%p) " REDO_F,
+		       rf->rf_m, REDO_P(&F(redo)));
+		M0_CO_YIELD_RC(CO(fom), m0_be_op_tick_ret(&F(reply_op),
+							  fom, RFS_WAITING));
+		m0_be_op_reset(&F(reply_op));
+	} while (F(next));
+
+	m0_be_op_fini(&F(reply_op));
+
+	recovery_machine_log_iter_fini(rf->rf_m, &rf->rf_log_iter);
+	M0_SET0(&rf->rf_log_iter);
+
+	M0_CO_FUN(CO(fom), heq_await(fom, out, eoq));
+}
+
+static void remote_recovery_fom_coro(struct m0_fom *fom)
+{
+	struct recovery_fom *rf = M0_AMB(rf, fom, rf_base);
+
+	M0_CO_REENTER(CO(fom),
+		      struct m0_be_op op;
+		      bool            eoq;
+		      enum m0_ha_obj_state state;
+		      void (*action)(struct m0_fom *fom,
+				     enum m0_ha_obj_state *out, bool *eoq);
+		      );
+
+	F(eoq) = false;
+	F(state) = M0_NC_UNKNOWN;
+
+	while (!F(eoq)) {
+		M0_LOG(M0_DEBUG, "remote recovery fom=%p, svc_fid=" FID_F ", "
+		       " proc_fid=" FID_F " handles %s state.", fom,
+		       FID_P(&rf->rf_tgt_svc), FID_P(&rf->rf_tgt_proc),
+		       m0_ha_state2str(F(state)));
+
+		switch (F(state)) {
+		case M0_NC_DTM_RECOVERING:
+			F(action) = rf->rf_is_volatile ? heq_await : restore;
+			break;
+		case M0_NC_FAILED:
+			if (ALL2ALL)
+				M0_LOG(M0_WARN, "Eviction is not supported.");
+			else
+				M0_IMPOSSIBLE("Eviction is not supported.");
```

Comment:
If this is intentional fall through, please add comments.

Reply:
3:Working on it -- adding comment in the code.

Reply:
Fixed/merged.